### PR TITLE
optimize lanczos call

### PR DESCRIPTION
**Pull request by esheldon**

### CODE REVIEW

**Comment by esheldon** [comment]
lanczos is now sub-dominant to `CloughTocher2DInterpolator`
```
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       37   63.892    1.727   64.169    1.734 _wcs_utils.py:50(__init__)
       74   49.850    0.674   50.053    0.676 _wcs_utils.py:55(__call__)
      665   43.336    0.065   43.336    0.065 {method 'normal' of 'mtrand.RandomState' objects}
     1926   42.082    0.022   42.461    0.022 interpolate.py:6(_interp_image)
       37   31.311    0.846   31.311    0.846 _lanczos.py:77(lanczos_resample_two)
        6   31.162    5.194   31.162    5.194 {method 'create_image_hdu' of '_fitsio.FITS' objects}
       76   11.569    0.152   63.165    0.831 memmappednoise.py:21(__init__)
      155   10.038    0.065   10.043    0.065 fitslib.py:330(__init__)
       76    9.125    0.120    9.125    0.120 {method 'flush' of 'mmap.mmap' objects}
       32    5.419    0.169    5.419    0.169 {method 'read_image' of '_fitsio.FITS' objects}
```

**Comment by esheldon** [comment]
The `_wcs_utils` stuff and the `_interp_image` are `CloughTocher2DInterpolator`, I wonder if we can improve that.

I'm surprised the random numbers are so slow, we should check that the rng is not being called extra times or something.

File: pizza_cutter/coadding/_lanczos.py [hunk 1]
```diff
@@ -58,12 +59,93 @@ def lanczos_resample(im, rows, cols, a=3):
...
+
+                val += im[y_pix, x_pix] * kernel
+
         res[i] = val
 
     return res
+
+
+@njit
+def lanczos_resample_two(im1, im2, rows, cols, a=3):
```
Comment (beckermr) [inline pizza_cutter/coadding/_lanczos.py]: We should add tests of this function.

File: pizza_cutter/coadding/_lanczos.py [hunk 2]
```diff
@@ -58,12 +59,93 @@ def lanczos_resample(im, rows, cols, a=3):
...
+    """Lanczos resample an image at the input row and column positions.
+
+    Parameters
+    ----------
+    im : np.ndarray
+        A two-dimensional array with the image values.
```
Comment (beckermr) [inline pizza_cutter/coadding/_lanczos.py]: Doc string needs adjusting.

File: pizza_cutter/coadding/_lanczos.py [hunk 3]
```diff
@@ -58,12 +59,93 @@ def lanczos_resample(im, rows, cols, a=3):
...
+    a : int, optional
+        The size of the Lanczos kernel. The default of 3 is a good choice
+        for many applications.
+
+    Returns
+    -------
+    values : np.ndarray
+        The resampled value for each row, column pair. Points whose
+        interpolation kernal does not touch any part of the grid are
+        returned as NaN.
```
Comment (beckermr) [inline pizza_cutter/coadding/_lanczos.py]: Here too.

File: pizza_cutter/coadding/_lanczos.py (hunk 1 above)
Comment (esheldon) [reply pizza_cutter/coadding/_lanczos.py]: it is getting tested through the `resample(` call, is that enough?

Comment (beckermr) [reply pizza_cutter/coadding/_lanczos.py]: Good point. LGTM.

**Comment by esheldon** [comment]
I added using a spline but this might not work well for pixmappy, we will see.

the current bottleneck is the interp code which is still using `CloughTocher2DInterpolator` because of warnings it was throwing, but almost equal is the random number generation.

@beckermr your cacheing didn't seem to reduce the fraction of the time spent in `normal()`
```
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     3721   81.152    0.022   81.894    0.022 interpolate.py:9(_interp_image_CT)
     1167   75.486    0.065   75.486    0.065 {method 'normal' of 'mtrand.RandomState' objects}
       75   64.306    0.857   64.306    0.857 _lanczos.py:77(lanczos_resample_two)
        6   30.050    5.008   30.050    5.008 {method 'create_image_hdu' of '_fitsio.FITS' objects}
     1952   22.095    0.011   22.095    0.011 {method 'read' of '_io.FileIO' objects}
      132   20.419    0.155  109.151    0.827 memmappednoise.py:21(__init__)

```

**Comment by beckermr** [comment]
Ack. There is a bug in how I was seeding the slices. This is why it is not working. Will fix but gotta run.

**Comment by esheldon** [comment]
I haven't found any alternatives from scipy for `CloughTocher2DInterpolator` for the masked pixel interpolations.

**Comment by esheldon** [comment]
A thought:  there is no reason to construct the interpolator from all good points in the image, we could use regions near the bad points to save time

**Comment by beckermr** [comment]
I am already doing this Last optimization. 

I work in patches of pixels and skip ones where no interpolation is needed.

**Comment by esheldon** [comment]
looking pretty good, with random numbers far down the list
```
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     2874   65.068    0.023   65.654    0.023 interpolate.py:9(_interp_image_CT)
       67   57.250    0.854   57.250    0.854 _lanczos.py:77(lanczos_resample_two)
        6   32.401    5.400   32.401    5.400 {method 'create_image_hdu' of '_fitsio.FITS' objects}
      175   14.178    0.081   14.183    0.081 fitslib.py:330(__init__)
      134   13.451    0.100   13.451    0.100 fitpack2.py:1035(__init__)
        4   11.920    2.980   11.920    2.980 {method 'execute' of 'cx_Oracle.Cursor' objects}
       52    9.164    0.176    9.164    0.176 {method 'read_image' of '_fitsio.FITS' objects}
      268    8.756    0.033    8.758    0.033 fitpack2.py:798(__call__)
      176    7.592    0.043    7.592    0.043 {method 'normal' of 'mtrand.RandomState' objects}

```

**Comment by esheldon** [comment]
I'm going to merge this.  The point of the PR was for the lanczos optimization which is done.  We can probably further optimize the `CloughTocher2DInterpolator` stuff but it can be in a different PR

**Pull request merged**